File: java/app/src/main/java/com/elisvobs/weather/ChangeCityController.java
package com.elisvobs.weather;

import android.content.Intent;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.google.android.material.textfield.TextInputEditText;

public class ChangeCityController extends AppCompatActivity {

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.city_changer);

        final TextInputEditText queryCity = findViewById(R.id.cityQuery);

        queryCity.setOnEditorActionListener(new TextView.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView textView, int i, KeyEvent keyEvent) {
                String newCity = queryCity.getEditableText().toString();
                Intent intent = new Intent(ChangeCityController.this, WeatherController.class);
                intent.putExtra("City", newCity);
                ChangeCityController.this.setResult(-1, intent);
                ChangeCityController.this.finish();
                return true;
            }
        });
    }

    public void backClick(View view) {
        // Go back and destroy the ChangeCityController
        finish();
    }
}

File: java/app/src/main/java/com/elisvobs/weather/ChangeCityController.kt
package com.elisvobs.weather

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.KeyEvent
import android.view.View
import android.widget.TextView

import androidx.appcompat.app.AppCompatActivity

import com.google.android.material.textfield.TextInputEditText

class ChangeCityController : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.city_changer)

        val queryCity = findViewById<TextInputEditText>(R.id.cityQuery)

        queryCity.setOnEditorActionListener { textView, i, keyEvent ->
            val newCity = queryCity.editableText.toString()
            val newCityIntent = Intent(this@ChangeCityController, WeatherController::class.java)

            // Adds what was entered in the EditText as an extra to the intent.
            newCityIntent.putExtra("City", newCity)

            // We started this activity for a result, so now we are setting the result.
            setResult(Activity.RESULT_OK, newCityIntent)

            // This destroys the ChangeCityController.
            finish()
            true
        }
    }

    fun backClick(view: View) {
        // Go back and destroy the ChangeCityController
        finish()
    }
}